src/ngAnimate/animateQueue.js

```diff
@@ -59,11 +82,24 @@ var $$AnimateQueueProvider = ['$animateProvider', function($animateProvider) {
   });
 
   rules.cancel.push(function(element, newAnimation, currentAnimation) {
-    var nO = newAnimation.options;
-    var cO = currentAnimation.options;
 
-    // if the exact same CSS class is added/removed then it's safe to cancel it
-    return (nO.addClass && nO.addClass === cO.removeClass) || (nO.removeClass && nO.removeClass === cO.addClass);
+
+    var nA = newAnimation.options.addClass;
+    var nR = newAnimation.options.removeClass;
+    var cA = currentAnimation.options.addClass;
+    var cR = currentAnimation.options.removeClass;
+
+    // early detection to save the global CPU shortage :)
+    if ((!isDefined(nA) && !isDefined(nR)) || (!isDefined(cA) && !isDefined(cR))) {
+      return false;
+    }
+
+    var cancelSomething = false;
+
+    cA = makeTruthyCssClassMap(cA);
+    cR = makeTruthyCssClassMap(cR);
+
+    return (hasMatchingClasses(nA, cR)) || (hasMatchingClasses(nR, cA));
```

Comment:
Sorry to be a pain but I just realised that we could even move the call to `makeTruthyCssClassMap` inside the `hasMatchingClasses` function too.